### PR TITLE
Added URI validation for disk creation source parameter

### PR DESCRIPTION
**Pull request by xiangyan99**
## What does this PR do?
`[Provide a clear, concise description of the changes]`

Added URI validation for disk creation source parameter

`[Add additional context, screenshots, or information that helps reviewers]`

## GitHub issue number?
`[Link to the GitHub issue this PR addresses]`

## Pre-merge Checklist
- [ ] Required for All PRs
    - [ ] **Read [contribution guidelines](https://github.com/microsoft/mcp/blob/main/CONTRIBUTING.md)**
    - [ ] PR title clearly describes the change
    - [ ] Commit history is clean with descriptive messages ([cleanup guide](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md))
    - [ ] Added comprehensive tests for new/modified functionality
    - [ ] Created a changelog entry if the change falls among the following: new feature, bug fix, UI/UX update, breaking change, or updated dependencies. Follow [the changelog entry guide](https://github.com/microsoft/mcp/blob/main/docs/changelog-entries.md)
- [ ] For MCP tool changes:
    - [ ] **One tool per PR**: This PR adds or modifies only one MCP tool for faster review cycles
    - [ ] Updated `servers/Azure.Mcp.Server/README.md` and/or `servers/Fabric.Mcp.Server/README.md` documentation
    - [ ] Validate `README.md` changes running the script `./eng/scripts/Process-PackageReadMe.ps1`. See [Package README](https://github.com/microsoft/mcp/blob/main/CONTRIBUTING.md#package-readme)
    - [ ] For new or modified tool descriptions, ran [`ToolDescriptionEvaluator`](https://github.com/microsoft/mcp/blob/main/eng/tools/ToolDescriptionEvaluator/Quickstart.md) and obtained a score of `0.4` or more and a top 3 ranking for all related test prompts
    - [ ] For tools with new names, including new tools or renamed tools, update [`consolidated-tools.json`](https://github.com/microsoft/mcp/blob/main/servers/Azure.Mcp.Server/src/Resources/consolidated-tools.json)
    - [ ] For **renamed** tools, follow the [Tool Rename Checklist](https://github.com/microsoft/mcp/blob/main/docs/tool-rename-checklist.md) and tag the PR with the `breaking-change` label
    - [ ] For new tools associated with Azure services or publicly available tools/APIs/products, add URL to documentation in the PR description
- [ ] Extra steps for **Azure MCP Server** tool changes:
    - [ ] Updated command list in [`servers/Azure.Mcp.Server/docs/azmcp-commands.md`](https://github.com/microsoft/mcp/blob/main/servers/Azure.Mcp.Server/docs/azmcp-commands.md)
    - [ ] Ran `./eng/scripts/Update-AzCommandsMetadata.ps1` to update tool metadata in [`azmcp-commands.md`](https://github.com/microsoft/mcp/blob/main/servers/Azure.Mcp.Server/docs/azmcp-commands.md) (required for CI)
    - [ ] Updated test prompts in [`servers/Azure.Mcp.Server/docs/e2eTestPrompts.md`](https://github.com/microsoft/mcp/blob/main/servers/Azure.Mcp.Server/docs/e2eTestPrompts.md)
    - [ ] 👉 For Community (non-Microsoft team member) PRs:
        - [ ] **Security review**: Reviewed code for security vulnerabilities, malicious code, or suspicious activities before running tests (`crypto mining, spam, data exfiltration, etc.`)
        - [ ] **Manual tests run**: added comment `/azp run mcp - pullrequest - live` to run *Live Test Pipeline*

### CODE REVIEW

File: tools/Azure.Mcp.Tools.Compute/src/Services/ComputeService.cs [hunk 1]
```diff
@@ -1750,6 +1758,29 @@ private static DiskCreationData CreateDiskCreationData(string? source, string? g
         };
     }
 
+    internal static readonly string[] s_allowedBlobHostSuffixes =
```
Comment (jongio) [inline tools/Azure.Mcp.Tools.Compute/src/Services/ComputeService.cs:1761]: This creates a third source of truth for blob storage domain suffixes - `StorageService.GetBlobEndpoint()` and `EndpointValidator.GetAllowedDomainSuffixes()` already maintain these.

Consider adding a `"storage-blob"` service type to `EndpointValidator` and calling `ValidateAzureServiceEndpoint(source, "storage-blob", armEnvironment)` instead. That validator already handles HTTPS enforcement, cloud-aware suffix selection via `ArmEnvironment` (so a Public Cloud user can't pass a China Cloud URI), and proper subdomain boundary checks. You'd need to pass `ArmEnvironment` into `CreateDiskCreationData`, which means threading it from the tenant service through `CreateDiskAsync` - a bit more plumbing but it aligns with how ACR and AppConfig handle the same pattern.

If keeping the standalone approach for now, make this array `private` - `readonly` on `string[]` only protects the reference, not the elements. Any internal caller could mutate the contents at runtime.

File: tools/Azure.Mcp.Tools.Compute/src/Services/ComputeService.cs [hunk 2]
```diff
@@ -1733,13 +1733,21 @@ private static DiskCreationData CreateDiskCreationData(string? source, string? g
...
+
+            var uri = new Uri(source);
+            ValidateAzureBlobStorageUri(uri);
+
             return new DiskCreationData(DiskCreateOption.Import)
             {
-                SourceUri = new Uri(source)
+                SourceUri = uri
             };
         }
```
Comment (jongio) [inline tools/Azure.Mcp.Tools.Compute/src/Services/ComputeService.cs:1752]: The flow here is hard to follow - the outer `if` enters for both http:// and https://, then immediately rejects http:// inside. A reader seeing the outer condition might think http is supported. Consider splitting into two clean branches:

```suggestion
        // HTTPS blob URIs - validate and import
        if (source.StartsWith("https://", StringComparison.OrdinalIgnoreCase))
        {
            var uri = new Uri(source);
            ValidateAzureBlobStorageUri(uri);

            return new DiskCreationData(DiskCreateOption.Import)
            {
                SourceUri = uri
            };
        }

        // Reject HTTP explicitly
        if (source.StartsWith("http://", StringComparison.OrdinalIgnoreCase))
        {
            throw new ArgumentException("Source URI must use HTTPS. HTTP endpoints are not allowed.", nameof(source));
        }
```

File: tools/Azure.Mcp.Tools.Compute/tests/Azure.Mcp.Tools.Compute.UnitTests/Disk/DiskCreateCommandTests.cs [hunk 3]
```diff
@@ -1219,4 +1219,150 @@ await _computeService.Received(1).CreateDiskAsync(
...
+        var nonAzureHttpUri = new Uri("http://evil-server.com/vhds/mydisk.vhd");
+
+        // Act & Assert
+        var ex = Assert.Throws<ArgumentException>(() =>
+            ComputeService.ValidateAzureBlobStorageUri(nonAzureHttpUri));
+        Assert.Contains("Azure Blob Storage endpoint", ex.Message);
+    }
+
+    [Theory]
+    [InlineData("https://evil-server.com/vhds/mydisk.vhd")]
```
Comment (jongio) [inline tools/Azure.Mcp.Tools.Compute/tests/Azure.Mcp.Tools.Compute.UnitTests/Disk/DiskCreateCommandTests.cs:1239]: Good coverage of the common attack vectors. A few more SSRF-relevant edge cases worth adding:
- `"https://account.blob.core.windows.net:4443/vhds/disk.vhd"` (non-default port)
- `"https://user:pass@account.blob.core.windows.net/vhds/disk.vhd"` (credentials in URI)

And for the positive test cases, a SAS token URI is the most common real-world import pattern:
- `"https://acct.blob.core.windows.net/vhds/disk.vhd?sv=2021-06-08&ss=b&srt=o&sp=r"`

**Review by jongio — CHANGES_REQUESTED** [review]
Good security fix - adding blob URI validation to disk creation closes an SSRF gap. The core logic is right (HTTPS enforcement + host allowlisting), but the implementation should build on the existing `EndpointValidator` pattern rather than creating a parallel validation path. See inline comments for specifics.

File: tools/Azure.Mcp.Tools.Compute/tests/Azure.Mcp.Tools.Compute.UnitTests/Disk/DiskCreateCommandTests.cs (hunk 3 above)
Comment (KarishmaGhiya) [reply tools/Azure.Mcp.Tools.Compute/tests/Azure.Mcp.Tools.Compute.UnitTests/Disk/DiskCreateCommandTests.cs:1239]: @xiangyan99 maybe you can address these feedbacks on main and open a PR for backporting as well?

Comment (xiangyan99) [reply tools/Azure.Mcp.Tools.Compute/tests/Azure.Mcp.Tools.Compute.UnitTests/Disk/DiskCreateCommandTests.cs:1239]: Yes. It is addressed in https://github.com/microsoft/mcp/pull/2337 which has been merged into main.
